### ISSUE TITLE
Decouple boosting types

### ISSUE THREAD

**Issue by candalfigomoro** [microsoft/LightGBM#3128]
With xgboost you can build (a sort of) Random Forest by setting `num_parallel_tree`>1 and `nrounds`=1

In LightGBM we can build (a sort of) Random Forest by setting `boosting`='rf'

Since the `num_parallel_tree` and `nrounds` params are decoupled in xgboost, what you can do is to set `num_parallel_tree`>1 and `nrounds`>1 to build a "**boosted random forest**". This is not possible in LightGBM afaik.

The fact that boosting types are mutually exclusive, make also impossible to create other combinations such as dart+goss (see https://github.com/microsoft/LightGBM/issues/2991).

The "rf" and "goss" modes should be decoupled from the boosting type. Maybe I want to use dart+rf or gbrt+rf (like in xgboost) to build boosted random forests. Maybe I want to use dart+goss.

@guolinke Maybe this can be considered for LightGBM 3 (https://github.com/microsoft/LightGBM/pull/3071)?

**Comment by guolinke** [comment]
I think it is a useful feature, but not trivial to implement it.
we will need to refactor the whole `boosting` part. And I don't have much time recently.
@shiyu1994 is helping me for the lightgbm project recently, but he is also busy this month. 
We can start refactoring in the next couple of months.

**Comment by candalfigomoro** [comment]
I think this would be a huge win in the long term because the code would be much more modular, but I understand the required effort is big

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by yiwiz-sai** [comment]
This feature is very useful, hope lightgbm team can support it.

`num_parallel_tree` is important to avoid overfitting, I think that's an advantage of xgboost,

**Comment by StrikerRUS** [comment]
Active work happens in #4827.
